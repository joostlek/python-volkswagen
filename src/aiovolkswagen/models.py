"""Models for Volkswagen."""
from dataclasses import dataclass

from mashumaro.mixins.orjson import DataClassORJSONMixin


@dataclass
class OpenIDConfiguration(DataClassORJSONMixin):

    issuer: str
    authorization_endpoint: str
    token_endpoint: str
    userinfo_endpoint: str
    jwks_uri: str
    scopes_supported: list[str]
    response_types_supported: list[str]
    grant_types_supported: list[str]
    acr_values_supported: list[str]
    subject_types_supported: list[str]
    id_token_signing_alg_values_supported: list[str]
    token_endpoint_auth_methods_supported: list[str]
    claims_supported: list[str]
    ui_locales_supported: list[str]
    revocation_endpoint: str
    code_challenge_methods_supported: list[str]
    end_session_endpoint: str

