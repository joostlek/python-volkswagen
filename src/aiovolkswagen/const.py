"""Constants for Volkswagen We Connect."""

X_APP_VERSION = "1.4.0"
X_APP_NAME = "SEATConnect"
USER_AGENT = "okhttp/3.10.0"
APP_URI = "cupraconnect://identity-kit/login"

CLIENT_ID = "30e33736-c537-4c72-ab60-74a7b92cfe83@apps_vw-dilab_com"

HEADERS_AUTH = {
    "Content-Type": "application/x-www-form-urlencoded",
    "User-Agent": USER_AGENT,
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
    "x-requested-with": X_APP_NAME,
    "Accept-Encoding": "gzip, deflate",
    "Connection": "keep-alive",
}
