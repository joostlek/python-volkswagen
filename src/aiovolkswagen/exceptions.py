"""Asynchronous Python client for Volkswagen."""


class VolkswagenError(Exception):
    """Generic exception."""


class VolkswagenConnectionError(VolkswagenError):
    """Volkswagen connection exception."""


class VolkswagenAuthenticationError(VolkswagenError):
    """Volkswagen authentication exception."""


class VolkswagenValidationError(VolkswagenError):
    """Volkswagen validation exception."""


class VolkswagenNotFoundError(VolkswagenError):
    """Volkswagen not found exception."""


class VolkswagenBadRequestError(VolkswagenError):
    """Volkswagen bad request exception."""
