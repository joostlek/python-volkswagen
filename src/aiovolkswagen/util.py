"""Utility functions for the aiovolkswagen library."""

import hashlib
import secrets
import string
from base64 import b64encode


def get_random_string() -> str:
    """Generate a random string."""
    chars = string.ascii_letters + string.digits
    text = "".join(secrets.choice(chars) for _ in range(10))
    sha256 = hashlib.sha256()
    sha256.update(text.encode())
    return b64encode(sha256.digest()).decode("utf-8")[:-1]
