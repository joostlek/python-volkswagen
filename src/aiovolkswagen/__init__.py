"""Asynchronous Python client for Volkswagen."""


from aiovolkswagen.exceptions import VolkswagenError, VolkswagenConnectionError, VolkswagenAuthenticationError, \
    VolkswagenValidationError, VolkswagenNotFoundError, VolkswagenBadRequestError
from aiovolkswagen.volkswagen import Volkswagen

__all__ = [
    "Volkswagen",
    "VolkswagenError",
    "VolkswagenConnectionError",
    "VolkswagenAuthenticationError",
    "VolkswagenValidationError",
    "VolkswagenNotFoundError",
    "VolkswagenBadRequestError",
]
