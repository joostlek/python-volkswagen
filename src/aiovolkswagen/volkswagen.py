"""Asynchronous Python client for Volkswagen."""

from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass
from importlib import metadata
from typing import Any, Self

from aiohttp import ClientSession
from aiohttp.hdrs import METH_GET

from aiovolkswagen.exceptions import VolkswagenConnectionError, VolkswagenError
from aiovolkswagen.models import OpenIDConfiguration
from aiovolkswagen.const import HEADERS_AUTH, APP_URI, CLIENT_ID
from aiovolkswagen.exceptions import (
    VolkswagenAuthenticationError,
    VolkswagenAccountLockedError,
    VolkswagenEULAError,
)
from aiovolkswagen.util import get_random_string
from aiovolkswagen.auth.login import (
    EmailFormParser,
    ClassicPasswordFormParser,
    DynamicPasswordFormParser,
)

from yarl import URL

VERSION = metadata.version(__package__)

LOGGER = logging.getLogger(__package__)


@dataclass
class Volkswagen:
    """Main class for handling connections with Volkswagen."""

    session: ClientSession | None = None
    request_timeout: int = 10
    _close_session: bool = False

    async def _request(
        self,
        url: str | URL,
        *,
        data: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> str:
        """Handle a request to Volkswagen."""
        headers = {
            "User-Agent": f"AioVolkswagen/{VERSION}",
            "Accept": "application/json",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        kwargs = {
            "headers": headers,
            "params": params,
            "json": data,
            **kwargs,
        }

        try:
            async with asyncio.timeout(self.request_timeout):
                response = await self.session.request(METH_GET, url, **kwargs)
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to Volkswagen"
            raise VolkswagenConnectionError(msg) from exception

        if response.status != 200:
            text = await response.text()
            msg = "Cannot connect to Volkswagen"
            raise VolkswagenConnectionError(
                msg,
                {"response": text},
            )

        content_type = response.headers.get("Content-Type", "")

        if "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected response from Volkswagen"
            raise VolkswagenError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        return await response.text()

    async def get_openid_configuration(self) -> OpenIDConfiguration:
        """Get the OpenID configuration."""
        response = await self._request(
            "https://identity.vwgroup.io/.well-known/openid-configuration"
        )
        return OpenIDConfiguration.from_json(response)

    async def login(self, username: str, password: str) -> None:  # pylint: disable=too-many-branches,too-many-statements
        """Login to Volkswagen."""
        openid_configuration = await self.get_openid_configuration()
        url = URL(openid_configuration.authorization_endpoint).with_query(
            {
                "redirect_uri": "cupraconnect://identity-kit/login",
                "nonce": get_random_string(),
                "state": get_random_string(),
                "response_type": "code id_token token",
                "client_id": CLIENT_ID,
                "scope": "openid profile address phone email birthdate nationalIdentifier cars mbb dealers badge nationality",
            }
        )
        if self.session is None:
            self.session = ClientSession()
            self._close_session = True
        request = await self.session.get(
            url, headers=HEADERS_AUTH, allow_redirects=False
        )
        if "Location" not in request.headers:
            raise VolkswagenAuthenticationError("Missing `location` header")
        location = request.headers["Location"]
        if "error" in location:
            location_url = URL(location)
            error = location_url.query["error"][0]
            if "error_description" in location:
                error = location_url.query["error_description"][0]
            raise VolkswagenAuthenticationError(error)

        if "signin-service" in location:
            location = await self._sign_in_to_service(
                location, openid_configuration, username, password
            )

        max_depth = 10
        try:
            while not location.startswith(APP_URI):
                if "error" in location:
                    location_url = URL(location)
                    error = location_url.query["error"][0]
                    if error == "login.error.throttled":
                        timeout = location_url.query["enableNextButtonAfterSeconds"][0]
                        raise VolkswagenAccountLockedError(
                            f"Account locked for {timeout} seconds"
                        )
                    if error == "login.errors.password_invalid":
                        raise VolkswagenAuthenticationError("Invalid password")
                    raise VolkswagenAuthenticationError(error)
                if "terms-and-conditions" in location:
                    raise VolkswagenEULAError("Accept terms and conditions")
                LOGGER.debug("Following redirect to %s", location)
                request = await self.session.get(
                    location, headers=HEADERS_AUTH, allow_redirects=False
                )
                if "Location" not in request.headers:
                    raise VolkswagenAuthenticationError("Missing `location` header")
                location = request.headers["Location"]
                max_depth -= 1
                if max_depth == 0:
                    raise VolkswagenAuthenticationError("Max redirect depth reached")
        except VolkswagenError as e:
            raise e
        except Exception as e:  # pylint: disable=broad-except
            if "code" not in location:
                raise VolkswagenAuthenticationError("Invalid code") from e
        location_url = URL(location)
        code = location_url.query["code"][0]
        id_token = location_url.query["id_token"][0]
        request = await self.session.post(
            "https://tokenrefreshservice.apps.emea.vwapps.io/exchangeAuthCode",
            headers=HEADERS_AUTH,
            data={"auth_code": code, "id_token": id_token, "brand": "cupra"},
        )
        if request.status != 200:
            if request.status >= 500:
                raise VolkswagenConnectionError("Invalid request")
            raise VolkswagenAuthenticationError("Invalid request")
        response = await request.json()
        LOGGER.info(response)

    async def _sign_in_to_service(
        self,
        location: str,
        openid_config: OpenIDConfiguration,
        username: str,
        password: str,
    ) -> str:
        """Sign in to the service."""
        assert self.session
        response = await self.session.get(
            location, headers=HEADERS_AUTH, allow_redirects=False
        )
        html = await response.text()
        parser = EmailFormParser()
        parser.feed(html)
        if not parser.action:
            raise VolkswagenAuthenticationError("No action found in form")
        headers = {
            **HEADERS_AUTH,
            "Referer": openid_config.authorization_endpoint,
            "Origin": openid_config.issuer,
        }
        form_data = {**parser.form_data, "email": username}
        response = await self.session.post(
            f"{openid_config.issuer}{parser.action}",
            headers=headers,
            data=form_data,
        )
        if response.status != 200:
            raise VolkswagenAuthenticationError("Invalid request")
        html = await response.text()
        post_action: str
        if "credentialsForm" in html:
            password_parser = ClassicPasswordFormParser()
            password_parser.feed(html)
            if not password_parser.action:
                raise VolkswagenAuthenticationError("No action found in form")
            data = {**password_parser.form_data, "password": password}
            post_action = password_parser.action
        elif "templateModel" in html:
            dynamic_password_parser = DynamicPasswordFormParser()
            dynamic_password_parser.feed(html)
            assert dynamic_password_parser.action
            data = {**dynamic_password_parser.form_data, "password": password}
            post_action = dynamic_password_parser.action
        else:
            raise VolkswagenAuthenticationError("No password form found")
        url = f"{openid_config.issuer}{post_action}"
        if "signin-service" not in url or CLIENT_ID not in url:
            url = f"{openid_config.issuer}/signin-service/v1/{CLIENT_ID}/{post_action}"
        headers["Referer"] = f"{openid_config.issuer}{parser.action}"
        response = await self.session.post(url, headers=headers, data=data)
        if response.status != 200:
            raise VolkswagenAuthenticationError("Invalid request")
        if "Location" not in response.headers:
            raise VolkswagenAuthenticationError("Missing `location` header")
        return response.headers["Location"]

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The Volkswagen object.
        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
