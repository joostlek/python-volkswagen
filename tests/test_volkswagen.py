"""Asynchronous Python client for Volkswagen."""

from __future__ import annotations

import asyncio
from datetime import date
from typing import TYPE_CHECKING, Any

import aiohttp
from aiohttp.hdrs import METH_GET, METH_POST, METH_PUT, METH_DELETE
from aioresponses import CallbackResult, aioresponses
import pytest
from yarl import URL

from aiovolkswagen.volkswagen import Volkswagen

from aiovolkswagen.exceptions import VolkswagenError, VolkswagenConnectionError
from tests import load_fixture

from .const import HEADERS

if TYPE_CHECKING:
    from syrupy import SnapshotAssertion


async def test_putting_in_own_session(
    responses: aioresponses,
) -> None:
    """Test putting in own session."""
    responses.get(
        "https://identity.vwgroup.io/.well-known/openid-configuration",
        status=200,
        body=load_fixture("openid-configuration.json"),
    )
    async with aiohttp.ClientSession() as session:
        vw = Volkswagen(session=session)
        await vw.get_openid_configuration()
        assert vw.session is not None
        assert not vw.session.closed
        await vw.close()
        assert not vw.session.closed


async def test_creating_own_session(
    responses: aioresponses,
) -> None:
    """Test creating own session."""
    responses.get(
        "https://identity.vwgroup.io/.well-known/openid-configuration",
        status=200,
        body=load_fixture("openid-configuration.json"),
    )
    vw = Volkswagen()
    await vw.get_openid_configuration()
    assert vw.session is not None
    assert not vw.session.closed
    await vw.close()
    assert vw.session.closed


async def test_unexpected_server_response(
    responses: aioresponses,
    volkswagen_client: Volkswagen,
) -> None:
    """Test handling unexpected response."""
    responses.get(
        "https://identity.vwgroup.io/.well-known/openid-configuration",
        status=400,
        headers={"Content-Type": "plain/text"},
        body="Yes",
    )
    with pytest.raises(VolkswagenError):
        assert await volkswagen_client.get_openid_configuration()



async def test_timeout(
    responses: aioresponses,
) -> None:
    """Test request timeout."""

    # Faking a timeout by sleeping
    async def response_handler(_: str, **_kwargs: Any) -> CallbackResult:
        """Response handler for this test."""
        await asyncio.sleep(2)
        return CallbackResult(body="Goodmorning!")

    responses.get(
        "https://identity.vwgroup.io/.well-known/openid-configuration",
        callback=response_handler,
    )
    async with Volkswagen(
        request_timeout=1
    ) as volkswagen_client:
        with pytest.raises(VolkswagenConnectionError):
            assert await volkswagen_client.get_openid_configuration()


async def test_about(
    responses: aioresponses,
    volkswagen_client: Volkswagen,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving openid configuration."""
    responses.get(
        "https://identity.vwgroup.io/.well-known/openid-configuration",
        status=200,
        body=load_fixture("openid-configuration.json"),
    )
    assert await volkswagen_client.get_openid_configuration() == snapshot

