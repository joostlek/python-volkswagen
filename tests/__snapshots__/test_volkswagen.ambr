# serializer version: 1
# name: test_about
  dict({
    'acr_values_supported': list([
      'https://identity.vwgroup.io/assurance/loa-2',
      'https://identity.vwgroup.io/assurance/loa-3',
    ]),
    'authorization_endpoint': 'https://identity.vwgroup.io/oidc/v1/authorize',
    'claims_supported': list([
      'iss',
      'sub',
      'aud',
      'iat',
      'exp',
      'nonce',
      'name',
      'email',
      'email_verified',
      'phone_number',
      'phone_number_verified',
      'updated_at',
    ]),
    'code_challenge_methods_supported': list([
      'S256',
    ]),
    'end_session_endpoint': 'https://identity.vwgroup.io/oidc/v1/logout',
    'grant_types_supported': list([
      'authorization_code',
      'refresh_token',
      'client_credentials',
      'password',
      'urn:ietf:params:oauth:grant-type:device_code',
      'urn:ietf:params:oauth:grant-type:token-exchange',
    ]),
    'id_token_signing_alg_values_supported': list([
      'RS256',
    ]),
    'issuer': 'https://identity.vwgroup.io',
    'jwks_uri': 'https://identity.vwgroup.io/v1/jwks',
    'response_types_supported': list([
      'code',
      'token',
      'id_token',
      'code token',
      'code id_token',
      'token id_token',
      'code token id_token',
    ]),
    'revocation_endpoint': 'https://identity.vwgroup.io/oidc/v1/revoke',
    'scopes_supported': list([
      'openid',
      'profile',
      'name',
      'badge',
      'preferred_language',
      'picture',
      'nickname',
      'gallery',
      'email',
      'birthdate',
      'address',
      'aboutMe',
      'userMileagePerYear',
      'phone',
      'vin',
      'healthReport',
      'nationality',
      'ecus',
      'wl_raw',
      'warning_lights',
      'profession',
      'nationalIdentifier',
      'driversLicense',
      'cars',
      'carConfigurations',
      'dealers',
      'mbb',
      'preferred_timezone',
      'parking_position',
      'lock_unlock',
      'doors_windows',
      'we_connect_vehicles',
      'battery',
      'errors',
      'intExtTemperature',
      'carMeasurements',
      'navigation',
      'vehicleOnlineCapabilities',
      'manageDestinations',
      'birthplace',
      'charging',
      'climatisation',
      'manageCharging',
      'manageClimatisation',
      'vehicleLights',
      'range',
      'mileage',
      'fuelLevels',
      'affinity',
      'digitalKey',
      'serviceScheduling',
      'tripStatistics',
      'rssFeeds',
      'messageCenter',
      'carComfortSetup',
      'serviceInterval',
      'assistantSystems',
      '12vBattery',
      'highVoltageBattery',
      'maintenance',
      'ecusMetaData',
      'climatisationProfile',
      'ignitionStatus',
      'plugAndCharge',
      'parkingBrakeStatus',
      'oilLevels',
      'beltLockStatus',
      'honk',
      'flashLights',
      'marketplace',
      'myPlaces',
      'gpsPosition',
      'vehicleSpeed',
      'skoda_connect_vehicles',
      'product_observation',
      'product_analysis',
      'verified_data',
      'theftalarm',
      'childPresenceAlert',
      'PrM_CDC_silent_mode',
      'Skoda_wishlist',
      'Skoda_leads',
      'remote_start',
      'gamloy',
    ]),
    'subject_types_supported': list([
      'public',
    ]),
    'token_endpoint': 'https://identity.vwgroup.io/oidc/v1/token',
    'token_endpoint_auth_methods_supported': list([
      'client_secret_basic',
      'client_secret_post',
    ]),
    'ui_locales_supported': list([
      'bg',
      'cs',
      'da',
      'de',
      'el',
      'en',
      'en-US',
      'en-GB',
      'es',
      'et',
      'fi',
      'fr',
      'hr',
      'hu',
      'is',
      'it',
      'ja',
      'ko',
      'lt',
      'lv',
      'mk',
      'nl',
      'no',
      'pl',
      'pt',
      'ro',
      'ru',
      'sk',
      'sl',
      'sq',
      'sr',
      'sv',
      'tr',
      'uk',
      'zh',
    ]),
    'userinfo_endpoint': 'https://identity-userinfo.vwgroup.io/oidc/userinfo',
  })
# ---
